port/osx/omrosdump.c

```diff
@@ -0,0 +1,519 @@
+		/* write each portion separately due to structure padding */
+		written = pwrite(corefile_fd, &threads[i].cmd, sizeof(uint32_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(uint32_t);
+		written = pwrite(corefile_fd, &threads[i].cmdsize, sizeof(uint32_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(uint32_t);
+		written = pwrite(corefile_fd, &threads[i].thread_state, sizeof(x86_thread_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_thread_state_t);
+		written = pwrite(corefile_fd, &threads[i].float_state, sizeof(x86_float_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_float_state_t);
+		written = pwrite(corefile_fd, &threads[i].exceptions, sizeof(x86_exception_state_t), file_off);
+		if (written < 0) {
+			perror("pwrite() error writing threads:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		file_off += sizeof(x86_exception_state_t);
+		mh64.sizeofcmds += threads[i].cmdsize;
+	}
+	mh64.ncmds += thread_count;
+
+	kr = list_segment_commands(task_port, &segments, &segment_count);
+	if (KERN_SUCCESS != kr) {
+		mach_error("error getting segment command data:\n", kr);
+		goto done;
+	}
+
+	seg_file_off = file_off + segment_count * sizeof(struct segment_command_64);
+	for (i = 0; i < segment_count; i++) {
+		vm_offset_t data_read = 0;
+		mach_msg_type_number_t data_size = 0;
+
+		kr = mach_vm_read(task_port, segments[i].vmaddr, segments[i].vmsize, &data_read, &data_size);
+		if (KERN_SUCCESS != kr) {
+			segments[i].cmd = 0;
+			unusable_segments += 1;
+			continue;
+		}
+
+		if ((0 == data_read) || (0 == data_size)) {
+			segments[i].cmd = 0;
+			unusable_segments += 1;
+			continue;
+		}
+
+		segments[i].fileoff = seg_file_off;
+
+		written = pwrite(corefile_fd, (void *)data_read, data_size, seg_file_off);
+		if (written < 0) {
+			perror("pwrite() error writing segment data:");
+			kr = KERN_FAILURE;
+			goto done;
+		}
+		mach_vm_deallocate(mach_task_self(), data_read, data_size);
+		seg_file_off += segments[i].vmsize;
+	}
+
+	/* update file offsets and shift written segments to cover the gap from unsuable segments */
+	for (i = 0; i < segment_count; i++) {
+		ssize_t read = 0;
+		ssize_t total_read = 0;
+		off_t seg_off_orig = segments[i].fileoff;
+		off_t seg_off_new = seg_off_orig - unusable_segments * sizeof(struct segment_command_64);
+		uint8_t buffer[4096];
+
+		if (LC_SEGMENT_64 != segments[i].cmd) {
+			continue;
+		}
+
+		while (total_read < segments[i].filesize) {
+			if (seg_off_orig + 4096 < segments[i].fileoff + segments[i].filesize) {
+				read = pread(corefile_fd, buffer, 4096, seg_off_orig);
```

Comment:
* 4096 should be replaced by `sizeof(buffer)` on these two lines
* parentheses should be added around both operands of `<`